### ISSUE TITLE
how to implement "Moto w/ Motion Token"

### ISSUE THREAD

**Issue by nanamma** [TencentARC/Moto#10]
thx for your hard work

Is there a configuration setting corresponding to the last two rows of Table 8 in the paper? My training results are close to those of "Moto w/o Motion Token." How can I implement the configuration for "Moto w/ Motion Token"? The code for the attention-mask design seems quite complex.

**Comment by nanamma** [comment]
Also, I found that the attention mask of part of latent motion in [`src/models/moto_gpt.py MotoGPT.forward()`](https://github.com/TencentARC/Moto/blob/main/moto_gpt/src/models/moto_gpt.py#L245) is not casual, which conflicts with `Next Latent Motion Token Prediction`.

**Comment by nanamma** [comment]
Regarding the training process, could you show the detailed training configuration (yaml config file) for each stage?
I really want to reproduce your training process. But my current reproduction results are not as well as yours.

**Comment by ChenYi99** [comment]
> Also, I found that the attention mask of part of latent motion in [`src/models/moto_gpt.py MotoGPT.forward()`](https://github.com/TencentARC/Moto/blob/main/moto_gpt/src/models/moto_gpt.py#L245) is not casual, which conflicts with `Next Latent Motion Token Prediction`.

The causal attention is implemented in the GPT2 code: [src/models/trajectory_gpt2.py#L340](https://github.com/TencentARC/Moto/blob/main/moto_gpt/src/models/trajectory_gpt2.py#L340). 
The causal attention mask is added to our attention mask in [src/models/trajectory_gpt2.py#L346](https://github.com/TencentARC/Moto/blob/main/moto_gpt/src/models/trajectory_gpt2.py#L346)

**Comment by ChenYi99** [comment]
> Regarding the training process, could you show the detailed training configuration (yaml config file) for each stage? I really want to reproduce your training process. But my current reproduction results are not as well as yours.

I have provided the detailed training configuration for each stage in [README](https://github.com/TencentARC/Moto/tree/main?tab=readme-ov-file#training). Did you strictly follow them to reproduce the results? Moreover, could you provide more details about your implementation?

**Comment by nanamma** [comment]
> > Regarding the training process, could you show the detailed training configuration (yaml config file) for each stage? I really want to reproduce your training process. But my current reproduction results are not as well as yours.
> 
> I have provided the detailed training configuration for each stage in [README](https://github.com/TencentARC/Moto/tree/main?tab=readme-ov-file#training). Did you strictly follow them to reproduce the results? Moreover, could you provide more details about your implementation?

I follow this https://github.com/TencentARC/Moto/issues/8#issuecomment-2673125981 and your paper to adjust the hyper parameters.

maybe can you provide the log of each stage? especially the loss. So that i can check whether my training of each stage is enough. thanks a lot

**Comment by ChenYi99** [comment]
[pretrain_moto_gpt_on_calvin.log](https://github.com/user-attachments/files/19769510/pretrain_moto_gpt_on_calvin.log)
[finetune_moto_gpt_on_calvin.log](https://github.com/user-attachments/files/19769509/finetune_moto_gpt_on_calvin.log)
[train_latent_motion_tokenizer_on_calvin.log](https://github.com/user-attachments/files/19769511/train_latent_motion_tokenizer_on_calvin.log)

**Comment by feng-tianli** [comment]
I followed the default configurations to train the model, but the results I obtained are far from those reported in the paper. As mentioned above, my results are also closer to those of "Moto w/o Motion Token." Did you manage to resolve this issue later? @nanamma

In addition, I noticed that the number of active codes during my training does not reach the levels mentioned in the training logs. The number of active codebook entries remains around 80–90 throughout training.

Could you please provide any insights or suggestions regarding these issues?

**Comment by nanamma** [comment]
> I followed the default configurations to train the model, but the results I obtained are far from those reported in the paper. As mentioned above, my results are also closer to those of "Moto w/o Motion Token." Did you manage to resolve this issue later? [@nanamma](https://github.com/nanamma)
> 
> In addition, I noticed that the number of active codes during my training does not reach the levels mentioned in the training logs. The number of active codebook entries remains around 80–90 throughout training.
> 
> Could you please provide any insights or suggestions regarding these issues?

My reproduction results for the first stage are quite good. The number of activated codes is close to what was reported in the original paper. And the losses are close, too.
It's best not to modify the training configuration, especially parameters like batch_size and gradient_accumulation_step. Although they can theoretically be adjusted proportionally, doing so may lead to some differences in results.
